Add gaps with forced ventilation

simonwacker opened the pull request:
Windows-CalcEngine can calculate the speed of the gas in the gap between layers caused by natural convection. When mechanical ventilation is used, the speed of the gas in the gap can be set to a constant value. This pull request adds the functionality of forced ventilation in the gap. In this case, the speed of the gas in the gap is simply set to a constant.

simonwacker committed:
Add gaps with forced ventilation

simonwacker committed:
Use existing `ForcedVentilation` member of `CGasLayer` declared in `LayerInterfaces`

simonwacker committed:
Set speed and direction for forced ventilation and calculate temperatures as for natural ventilation

simonwacker committed:
Assert that neither gap layer around a shade has forced ventilation

simonwacker commented on src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilation.unit.cpp:
```diff
@@ -0,0 +1,136 @@
+    ASSERT_TRUE(aLayer != nullptr);
+    auto airflowReferencePoint = aLayer->getAirflowReferencePoint(0.5);
+    EXPECT_NEAR(6911.7392493523666, airflowReferencePoint, 1e-4);
+}
```
@bbuenoun Do these tests and numbers make sense?

bbuenoun replied:
What is the inlet temperature of the gap: the indoor temperature or the outdoor temperature? in a case with only two layers, both possibilities are possible

simonwacker replied:
[The environment used to calculate in and out temperatures of a ventilated gap is the one adjacent to the shade layer adjacent to the gap](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/BaseShade.cpp#L168-L180). So, in the layering `outdoor - solid - gap - shade - indoor`, it is the indoor environment and in the layering `outdoor - shade - gap - solid - indoor`, it is the outdoor environment. For those setups, the temperatures are set on the lines https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/BaseShade.cpp#L334-L347 (Looking closer at this, I am unsure if I implemented it correctly for forced ventilation. In this implementation, `m_inTemperature` is set to `TgapOut` or `tempEnvironment` [depending on which of the two is greater](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/BaseShade.cpp#L334) and [depending on whether `forcedVentilation.VerticalDirection` is up or down](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/IGUVentilatedGapLayer.cpp#L75-L82).). Note that [`tempEnvironment`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/BaseShade.cpp#L303) is the gas temperature of the environment adjacent to the shade layer.

simonwacker replied:
The new numbers with the promised changes can be seen on the green lines in https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/7dc4422ceebcdeb37ac8a5cdca8220a1255693a1#diff-f1510538eafc05022ed4282de85aba1f6ce25356865aa7c328620229546650a7

bbuenoun replied:
The results of the winter unit tests are plausible

bbuenoun replied:
Please indicate the values of these variables in winter and summer unit tests:
m_HInput, m_HCoefficientModel, m_IRCalculatedOutside

simonwacker replied:
I added tests for those members in https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/3492a6d8f5966ba296555b618867719a9d401239 To do so, I had to add `get`-functions for those members to `CEnvironment` and to `CSystem`. In both cases, `m_HInput` and `m_IRCalculatedOutside` is `0` and `m_HCoefficientModel` is `BoundaryConditionsCoeffModel::CalculateH`.

simonwacker committed:
Test temperatures of solid and shade layers surrounding the forced ventilated gap layer

simonwacker committed:
Test forced ventilation with non-zero solar radiation

simonwacker committed:
Add forced ventilation with zero air speed

simonwacker commented on src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiation.unit.cpp:
```diff
@@ -0,0 +1,182 @@
+        auto airTemperature = 255.15;   // Kelvins
+        auto airSpeed = 5.5;            // meters per second
+        auto tSky = 255.15;             // Kelvins
+        auto solarRadiation = 700.0;
```
@bbuenoun In this test I set the solar radiation to `700`. The computed numbers though do **not** change compared to [the test with solar radiation `0`](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/files#diff-f1510538eafc05022ed4282de85aba1f6ce25356865aa7c328620229546650a7R20).

bbuenoun replied:
We have to check the (hidden) optical properties of the layers. The absorptivity might be set to zero.

simonwacker replied:
In https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/070c3b13645b285690d2f9fd627436226039b7fc I set the absorptance to a non-zero value, called `solve` on the Tarcog system in the set-up, and also tested radiosity values. Without calling `solve`, setting the absoptance didn't have an effect on the values. With calling `solve`, all values changed. I don't really know what solve does. In the code it looks like the system has an initial state and calling solve makes the system proceed to the next stable state. Its implementation is https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/NonLinearSolver.cpp#L54

simonwacker replied:
The new numbers with about absorptance 10% for the glass layer and absorptance 70% for the shade layer can be seen on the green lines in https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/7dc4422ceebcdeb37ac8a5cdca8220a1255693a1#diff-02f7c1e784833ccc1b166b36f337d7a5d836ed05759ab60fa86880260d995ef0

bbuenoun replied:
Please modify this unit test (let's call it "summer unit test") with the following boundary conditions:
Irradiation: 783 W/m2.
Outdoor temperature: 305.15 K.
Indoor temperature: 297.15 K
Please indicate what are the (hidden) solar reflectance and solar transmittance of the solid and shade layers, respectively.

simonwacker replied:
I added the requested test with summer temperatures and irradiation: [GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer.unit.cpp). I also specified transmittance and emissivity values explicitely in https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer.unit.cpp#L40-L44 and https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer.unit.cpp#L57-L61 and noted how these values are used to calculate the reflectance in [`Surface.cpp#calculateReflectance`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Surface.cpp#L115-L126). The implicit defaults used before are defined in [`Layers.hpp#solid`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Layers.hpp#L22-L27) and [`Layers.hpp#shading`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Layers.hpp#L34-L42).

bbuenoun replied:
You specified the infrared transmittance (solidLayerFrontIRTransmittance), but we need the SolarTransmittance (the variable might be called "solidLayerFrontSolarTransmittance"), but I'm not sure

simonwacker replied:
There are no other arguments to specify the transmittance that I could find. May it be that the argument given to [the helper function `solid`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Layers.cpp#L14-L26) as well as to [the constructor of solid layers](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/IGUSolidLayer.cpp#L41-L55) have the wrong name (I used the names of those arguments to name my variables in the tests)? In [the helper function `shading`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Layers.cpp#L63-L93) the arguments used in the very same way are named `frontTransmittance` and `backTransmittance` (these arguments are in the case that the shade layer is closed even passed to the helper function `solid` as `...IRTransmittance`; so at least here is an inconsistency in the usage). In the case that the shade layer is open, the transmittance arguments are in both usages passed to [the constructor of `CSurface`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/Surface.cpp#L28-L37) in which the argumet is named `t_Transmittance`.

bbuenoun replied:
Ok. We need to discuss with Simon Vadanovic how the shortwave radiation problem is solved in Wincalc and what is the meaning of "solidLayerSolarAbsorptance", whether it already takes into account what is transmitted from the first layer.

bbuenoun replied:
The results of [GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithSolarRadiationInSummer.unit.cpp) are plausible

simonwacker commented on src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp:
```diff
@@ -0,0 +1,182 @@
+        ASSERT_TRUE(shadeLayer != nullptr);
+
+        auto gapThickness = 0.0127;
+        auto gapAirSpeed = 0.0;
```
@bbuenoun In this test I set the gap air speed to `0`. Compared to [the test with the value `0.5`](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/files#diff-f1510538eafc05022ed4282de85aba1f6ce25356865aa7c328620229546650a7R67) only the gain energy changes from `-173.19394352133912` to `0.0`. The reason is that the way the computation currently works, we **force** an air speed of `0` no matter what in [`CIGUShadeLayer::calcEdgeShadeFlow`](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/files#diff-373f0b5a3bab54faaf37cfa2a370957a33143cf5af21a40aa39861baea9356e1R297-R300) (any possible natural ventilation is suppressed). And this air speed of `0` is [used as multiplicative factor in the calculation of `CIGUVentilatedGapLayer::ventilatedFlow`](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/src/IGUVentilatedGapLayer.cpp#L182) resulting in the ventilated flow `0`.

bbuenoun replied:
We have to compare the test with air_velocity=0 with a test without natural ventilation, i.e. sealed and narrow gap between two layers.

simonwacker replied:
In https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/4bda8f0a9a9848b4230f686345c518d9cd07ebdb I added a test with a naturally ventilated [narrow](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/4bda8f0a9a9848b4230f686345c518d9cd07ebdb#diff-77b34fc1601b2388bada4d0314de1afe488763836024690d16cc0a15b08d9f92R66) gap at the edge. The [gain flow](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/4bda8f0a9a9848b4230f686345c518d9cd07ebdb#diff-77b34fc1601b2388bada4d0314de1afe488763836024690d16cc0a15b08d9f92R116) is close to zero.

simonwacker replied:
I made a naive attempt to implement sealed forced-ventilation gaps in https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/156b5e035530a00945bf5bb65829df06d1fd703b I currently use the previous layer's temperature as inlet temperature for the gap. Should one be able to configure which layer is used for the inlet temperature? Or should it always be the previous or next layer? I added a test with a [normal sealed forced-ventilation gap](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/156b5e035530a00945bf5bb65829df06d1fd703b#diff-2bc71bc9c8ff5c69e24cd6a17dcce54e9424c2df4496d2976ae854b5a6354b4a), one [such that is narrow](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/156b5e035530a00945bf5bb65829df06d1fd703b#diff-f682c31f98199791424b8439fff92ca8201334d507c0f6f96b8d839cc7c143f2), and one [such that is narrow and has air speed zero](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/aca8734711b4c9721d6a919f5a92ea06fbe7dfd0). Which test should produce similar numbers as [GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp)? [GapLayerSealedForcedVentilationNarrow](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerSealedForcedVentilationNarrow.unit.cpp)?

bbuenoun replied:
[GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp) has to be compared with a configuration where all the gaps (Atop, Abottom, Afront, Aright, Aleft) are set to zero.

simonwacker replied:
As of https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/3e1e6a4c6507a941cef18a03726f90bfee6e8ea9 sealed forced-ventilation gaps are now modelled by putting a closed shade to one side of a forced-ventilation gap. For a code example see https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/3e1e6a4c6507a941cef18a03726f90bfee6e8ea9#diff-2bc71bc9c8ff5c69e24cd6a17dcce54e9424c2df4496d2976ae854b5a6354b4aR44-R46 and https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/3e1e6a4c6507a941cef18a03726f90bfee6e8ea9#diff-2bc71bc9c8ff5c69e24cd6a17dcce54e9424c2df4496d2976ae854b5a6354b4aR57. The gain energies of [GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp) and [GapLayerSealedForcedVentilationNarrow](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerSealedForcedVentilationNarrow.unit.cpp) differ only slightly. The temperatures however differ by about 3 to 5 degrees Celsisus. The gain energies of [GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp) compared to [GapLayerSealedForcedVentilation](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerSealedForcedVentilation.unit.cpp) differs by about 132. The temperatures also differ by 3 to 5 degrees Celsius.

bbuenoun replied:
Please compare [GapLayerAtEdgeForcedVentilationWithZeroAirSpeed](https://github.com/ise-bipv/Windows-CalcEngine/blob/add-gaps-with-forced-ventilation/src/Tarcog/tst/units/GapLayerAtEdgeForcedVentilationWithZeroAirSpeed.unit.cpp) with a configuration with the same geometry, naturally ventilated, where all the gaps (Atop, Abottom, Afront, Aright, Aleft) are set to zero.

simonwacker replied:
I [added a test for the requested configuration](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/5c6fdf145db1467262532c1768c4da5f1ddb8faa) (sorry for being so stubborn) and compared the numbers. The temperatures only differ in the third decimal place and the gain energy differs by `0.0182...`.

simonwacker reviewed (commented):
I implemented two other tests, [one with non-zero solar radiation](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/2cc2a8cc9105b252f545f38f4dd7803480ec84fe) and [another one with zero air speed](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/42ce33dad40053dd64fa8a2ad4f1750e6dbb4ffb). I also [added additional tests for the solid and shade layer temperatures](https://github.com/ise-bipv/Windows-CalcEngine/pull/2/commits/1f81793390f794c535a38813414707e789c236e0). As you expected, the front temperature of the gap is the back temperature of the solid layer and the back temperature of the gap is the front temperature of the shade layer.

simonwacker committed:
Call `solve` on the Tarcog system after setting it up, change expected values accordingly, and test also test radiosities

simonwacker committed:
Test the indoor heat flow for a system with a forced radiation gap positive solar radiation

simonwacker committed:
Test natural ventilation of gap at edge

simonwacker committed:
Test natural ventilation of narrow gap at edge

simonwacker committed:
Call `solve` on the Tarcog system after setting it up to calculate layer values like their temperatures, remove air-flow direction from forced ventilation as it does not matter, use the environment temperature as inlet temperature for at-edge gaps with forced ventilation and the gap temperature as out temperature, use about 10% absorptance for galsses and about 70% for shades, and adjust the expected values in tests accordingly

simonwacker committed:
Support and calculate temperatures of sealed forced-ventilation gaps, that is, forced-ventilation gaps sandwiched between solids

simonwacker committed:
Make scoped traces in tests more expressive

simonwacker committed:
Test sealed and narrow forced-ventilation gap layer with zero air speed

simonwacker committed:
Turn to-do into note and remove legacy questions

simonwacker commented on src/Tarcog/src/IGUVentilatedGapLayer.cpp:
```diff
@@ -147,10 +159,77 @@ namespace Tarcog
+            double tempPreviousLayer = t_PreviousLayer->getTemperature(Side::Back);
+            while(!converged)
+            {
+                double TavGap = averageTemperature();
+                // To calculate `beta` we do not invoke `betaCoeff` because
+                // doing so results in an infinite recursion as it calls
+                // `calculateLayerHeatFlow`, which calls
+                // `calculateConvectionOrConductionFlow`, which calls this
+                // method.
+                double beta = exp(-m_Height / characteristicHeight());
+                double alpha = 1 - beta;
```
Do these variables change during the loop? If not, extract them!

simonwacker committed:
Use closed shades to model sealed forced-ventilation gaps

simonwacker committed:
Make transmittance and emissivity explicit and tell that reflectance is equal to `1 - emissivity - transmittance (the implicit defaults are defined in `Layers.hpp#solid` and `Layers.hpp#shading`, and the calculation of reflectance is defined in `Surface.cpp#calculateReflectance` which is called by the constructors of `ISurface`)

simonwacker committed:
Put declarations of air-direction enumerations back to where they used to be (to have an uncluttered difference view)

simonwacker committed:
Add test for sealed gap at edge with natural ventilation

simonwacker committed:
Add getters for `m_HInput`, `m_HCoefficientModel`, `m_IRCalculatedOutside` to environment and to system and test those values in winter and summer tests of forced-ventilation gaps with solar radiation

simonwacker committed:
Delete ventilation tests with narrow gaps

simonwacker committed:
Improve naming of ventilated gap tests

simonwacker committed:
Add missing abbreviation IR for infrared to front and back transmittance argument names

simonwacker committed:
Name `closedShading` but `sealedLayer`

vidanovic committed:
ForcedVentilation structure removed since it contained only one member double

vidanovic committed:
Separation for setting of geometry vs setting direction.

vidanovic committed:
Flow geometry is set only once

vidanovic committed:
Inlet temperature now have to be set in the separate routine and can be updated within the iterations or just left constant.

vidanovic committed:
Outlet temperature for the airflow calculations is moved into ventilated gap class

vidanovic committed:
Setting flow temperatures is moved into ventilated gap class.

vidanovic committed:
Thermally driven speed of two connected gaps is moved into ventilated gap class.

vidanovic committed:
In between shade ventilated gap calculations are moved into ventilated gap class

vidanovic committed:
Initialization of forced ventilation air speed moved from calcEdgeShadeFlow (Shading layer class) into ventilated gaps class.

vidanovic committed:
Airflow reference point calculations are now private functions. This also required to remove some of the unit tests for these private functions.

vidanovic committed:
Moving more functions into private

vidanovic committed:
Calculation of airflow for the edge case is now completely moved out of shading layer class

vidanovic committed:
Thermally driven ventilation between two gaps is moved into Ventilated Gap class

vidanovic committed:
Couple of more functions from ventilated gap moved into the private section.

vidanovic committed:
Top and Bottom opening multipliers are no longer saved with the ventilation gap class but rather impendances are calculated only once and stored premanently. This will also speed up ventilation calculations.

vidanovic committed:
Fixed the issue in thermally driven airflow calculations where the inlet temperature was not updating correctly.

vidanovic committed:
Inlet and outlet temperatures are stored in a single structure.

vidanovic committed:
Forced airflow calculations are now called for non-shading layers in the system. Unit test for forced ventilation are not passing and results needs to be verfied now with the new approach.

vidanovic committed:
Update to unit test that perform forced ventilation of the gap at the edge (with outside and inside inlet airflow).

vidanovic committed:
Forced ventilation on the edge cases with the different values of the solar radiation.

vidanovic committed:
Update to all unit test performed. Still missing test for in between forced ventilation.

vidanovic committed:
Forced ventilation for in between case is now in tests as well.

simonwacker committed:
Merge pull request #4 from ise-bipv/airflow-refactoring

Airflow refactoring

simonwacker commented:
@christoph-maurer I merged Simon's refactorings. Just pull the latest version and then you can start improving and adding further examples in the form of tests.

simonwacker commented:
@vidanovic I reviewed your refactorings, really like them, and merged them into this feature branch. The changes add clarity, decouple functionalities and put them into the right place. Thanks! I didn't have the insight into the code base to be able to do the implementation like that.

christoph-maurer committed:
Copy example GapLayerAtEdgeForcedVentilationInsideAir.unit.cpp and rename it as the first step to create a new example

vidanovic commented:
@simon-wacker It does look good. I have also found very helpful things that you did too. It did help me to perform this refactoring. Few days before the vacation I was trying to make python equivalent of the algorithm, but my results were not matching. I might spend more time in trying to determine what is causing the difference once we are back to work (Tuesday, January 17th 2023).

christoph-maurer committed:
Rename the class `TestGapLayerAtEdgeForcedVentilationOutsideAir` to `TestGapBetweenExteriorShadingAndGlassForcedVentilationOutsideAir`. Test successfully.

christoph-maurer committed:
The results of the new test indicate that the shading has zero absorptance.

christoph-maurer committed:
Fail to add absorptance as in DoubleOutsidePerforatedShade_SHGC.unit.cpp . Error message: `[----------] 5 tests from TestGapBetweenIrradiatedExteriorShadingAndGlassForcedVentilationOutsideAir
[ RUN      ] TestGapBetweenIrradiatedExteriorShadingAndGlassForcedVentilationOutsideAir.GapLayerSurfaceIRFlow
unknown file: Failure
C++ exception with description "map::at" thrown in SetUp().
[  FAILED  ] TestGapBetweenIrradiatedExteriorShadingAndGlassForcedVentilationOutsideAir.GapLayerSurfaceIRFlow (0 ms)
[ RUN      ] TestGapBetweenIrradiatedExteriorShadingAndGlassForcedVentilationOutsideAir.GainEnergy
unknown file: Failure
C++ exception with description "map::at" thrown in SetUp().`

simonwacker committed:
Do not upgrade gap layers that already are ventilated

simonwacker committed:
Remove unused includes

simonwacker committed:
Use `CSingleSystem` as `m_TarcogSystem`

simonwacker committed:
Declare missing air temperature variable

simonwacker committed:
Assert non-nullness of shade layer

simonwacker committed:
Solve the system at end of test set-up

simonwacker committed:
Give variables more readable and expressive names

simonwacker committed:
Correct expected numbers

simonwacker committed:
Correct expected ventilated flow in `DoubleClarIndoorShadeAir` tests

christoph-maurer committed:
Merge pull request #6 from ise-bipv/fix-outside-air-test

Fix outside air test

christoph-maurer committed:
Create three test cases for validation with the software Winther

christoph-maurer committed:
Create tests for interior screen with varying height and mass flow of the forced convection.

christoph-maurer committed:
Rename the validation tests to make them more consistend with the other `GapBetween*` tests.

christoph-maurer committed:
Finish renaming the 8 tests for validation and explanation

christoph-maurer committed:
Use new lines consistently in the blocks of the 8 new tests

christoph-maurer committed:
Add test with exterior shading and forced convection with roomTemperature for comparison.

christoph-maurer committed:
Add tests to compare zero forced convection with the natural convection of a gap without forced convection. Why is there a difference? Relates to #5

christoph-maurer committed:
Start test with 2 glass panes and 1 shading in between, but the test do not run properly.

christoph-maurer committed:
Fix test GapBetweenIrradiatedGlassPanesWithIntegratedShadingForcedVentilation by adding `m_TarcogSystem->solve();`

simonwacker committed:
Add getters for inlet and outlet temperatures

christoph-maurer committed:
Merge pull request #7 from ise-bipv/add-getters-for-inlet-and-oulet-temperature

Add getters for inlet and outlet temperatures

christoph-maurer committed:
Add example with very thick glass to calculate the external and internal heat transfer coefficients

christoph-maurer committed:
Fail to print the indoor radiative heat transfer coefficient

Error message: /home/christoph/Documents/Windows-CalcEngine/src/Tarcog/tst/units/GapBetweenIrradiatedGlassPanesForcedVentilationInsideAirSummerValidation.unit.cpp:123:19: error: ‘CIndoorEnvironment’ does not name a type
  123 |     [[nodiscard]] CIndoorEnvironment * GetIndoorRadiativeHeatTransferCoefficient() const
      |                   ^~~~~~~~~~~~~~~~~~
/home/christoph/Documents/Windows-CalcEngine/src/Tarcog/tst/units/GapBetweenIrradiatedGlassPanesForcedVentilationInsideAirSummerValidation.unit.cpp:127:6: warning: extra ‘;’ [-Wpedantic]
  127 |     };
      |      ^
      |      -
/home/christoph/Documents/Windows-CalcEngine/src/Tarcog/tst/units/GapBetweenIrradiatedGlassPanesForcedVentilationInsideAirSummerValidation.unit.cpp: In member function ‘virtual void TestGapBetweenIrradiatedGlassPanesForcedVentilationInsideAirSummerValidation_IndoorHeatTransferCoefficients_Test::TestBody()’:
/home/christoph/Documents/Windows-CalcEngine/src/Tarcog/tst/units/GapBetweenIrradiatedGlassPanesForcedVentilationInsideAirSummerValidation.unit.cpp:216:51: error: ‘GetIndoorRadiativeHeatTransferCoefficient’ was not declared in this scope; did you mean ‘indoorRadiativeHeatTransferCoefficient’?
  216 |     auto indoorRadiativeHeatTransferCoefficient = GetIndoorRadiativeHeatTransferCoefficient();
      |                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                   indoorRadiativeHeatTransferCoefficient
make[1]: *** [src/Tarcog/CMakeFiles/Tarcog_tests.dir/build.make:776: src/Tarcog/CMakeFiles/Tarcog_tests.dir/tst/units/GapBetweenIrradiatedGlassPanesForcedVentilationInsideAirSummerValidation.unit.cpp.o] Error 1

christoph-maurer committed:
No error message anymore, but no results neither.

christoph-maurer committed:
The heat transfer coefficients of the summer validation look fine

christoph-maurer committed:
Test the heat transfer coefficients of all three validation cases

christoph-maurer committed:
Merge pull request #8 from ise-bipv/print-heat-transfer-coefficients

Test the heat transfer coefficients of the validation cases

christoph-maurer committed:
Describe when the  is used

christoph-maurer committed:
Add old test from Windows-CalcEngine

christoph-maurer committed:
Remove the tests which are replaced by the tests `GapBetween*`

simonwacker committed:
Correct expected values

simonwacker committed:
Add test with naturally ventilated and sealed gap to cross validate the forced ventilation test `TestGapBetweenIrradiatedGlassAndInteriorShadingZeroForcedVentilation`

simonwacker committed:
Merge pull request #5 from ise-bipv/more-examples-of-gaps-with-forced-ventilation

Create more examples of gaps with forced ventilation

simonwacker committed:
Revert commit https://github.com/ise-bipv/Windows-CalcEngine/commit/3492a6d8f5966ba296555b618867719a9d401239 the getters introduced there are not used anywhere and at least `getIRCalculatedOutside` is actually not needed due to https://github.com/ise-bipv/Windows-CalcEngine/commit/3492a6d8f5966ba296555b618867719a9d401239#commitcomment-105581336

christoph-maurer committed:
Compare with 1mm gap zeroForced and natural convection

christoph-maurer committed:
Update name of the tests

christoph-maurer committed:
The IRFlow differs between the 1mmGap tests

christoph-maurer committed:
The temperatures match with about +/- 0.02 K

christoph-maurer committed:
The IRFlow matches with +/- 0.02 W/m². The energy gain of ZeroForcedConvection is 0. The energy gain of NaturalConvection is 0.0439 W/m². Both is plausible.

simonwacker committed:
Correct expected values

simonwacker committed:
Merge pull request #9 from ise-bipv/compare-1mm-gap

Compare with 1mm gap zeroForced and natural convection

christoph-maurer commented:
@vidanovic In order to validate the implementation of the forced convection, we have compared natural convection and forced convection with zero air speed both with a gap thickness of 1 mm in #9. The validation was successful, the results of both tests are very similar.

vidanovic commented:
@christoph-maurer Great. I am going to take a quick look and merge it. Thanks!!!

vidanovic reviewed (approved):
Results do make sense to me. Unless someone else objects to this, I would merge this into the main and then merge it into the our main LBNL branch.

simonwacker commented:
> Results do make sense to me. Unless someone else objects to this, I would merge this into the main and then merge it into the our main LBNL branch.

There is the sibling merge request https://github.com/LBNL-ETA/Windows-CalcEngine/pull/144

simonwacker commented:
The sibling merge request https://github.com/LBNL-ETA/Windows-CalcEngine/pull/144 was merged into the official repository. :tada:
